example/tests/pin_api_test.py

```diff
@@ -0,0 +1,209 @@
+    p.data = 2
+  with pytest.raises(OSError):
+    p.drive(2)
+  with pytest.raises(OSError):
+    p.verify(2)
+  with pytest.raises(OSError):
+    p.set(2)
+  assert p.action == "HighZ"
+
+  # Check that pins are available on subblocks
+  assert len(origen.dut.pins) == 2
+  assert origen.dut.sub_blocks["core1"]
+  assert len(origen.dut.sub_blocks["core1"].pins) == 0
+  assert origen.dut.sub_blocks["core1"].add_pin("test_pin_core1")
+  assert len(origen.dut.sub_blocks["core1"].pins) == 1
+  p = origen.dut.sub_blocks["core1"].pin("test_pin_core1")
+  assert p._path == "core1"
+  assert len(origen.dut.pins) == 2
+
+  # Add a pin alias
+  p = origen.dut.pin('test_pin')
+  #p.add_alias('test_alias')
+  origen.dut.add_pin_alias('test_pin', 'test_alias1')
+  alias_p = origen.dut.pin('test_alias1')
+  assert alias_p.name == p.name
+  assert len(origen.dut.pins) == 2
+
+  # Add multiple aliases at once.
+  origen.dut.add_pin_alias('test_pin', 'test_alias2', 'test_alias3')
+  assert len(origen.dut.pins) == 2
+  assert origen.dut.pin('test_alias2').name == 'test_pin'
+  assert origen.dut.pin('test_alias3').name == 'test_pin'
+
+  # Error conditions related to aliasing
+  with pytest.raises(OSError):
+    # Pin doesn't exists
+    origen.dut.add_pin_alias("blah", "test_blah")
+  with pytest.raises(OSError):
+    # alias already exists as a pin
+    origen.dut.add_pin_alias("test_pin", "test_pin")
+  with pytest.raises(OSError):
+    # Alias already exists
+    origen.dut.add_pin_alias("test_pin", "test_alias1")
+  with pytest.raises(OSError):
+    # Ensure we cannot add a pin whose name conflicts with an existing alias.
+    origen.dut.add_pin("test_alias1")
+
+  # Check some of the dict-like API
+  assert "test_pin" in origen.dut.pins
+  assert "test_alias1" in origen.dut.pins
+  keys = origen.dut.pins.keys()
+  values = origen.dut.pins.values()
+  assert set(keys) == set(['test_pin', 'other_pin'])
+  assert len(values) == 2
+  assert isinstance(values[0], _origen.dut.pins.Pin)
+  for name in origen.dut.pins:
+    assert name in keys
+  d = dict(origen.dut.pins)
+  assert isinstance(d, dict)
+  assert isinstance(d["test_pin"], _origen.dut.pins.Pin)
+  assert isinstance(d["other_pin"], _origen.dut.pins.Pin)
+  assert len(d) == 2
+  for n, p in origen.dut.pins.items():
+    assert isinstance(n, str)
+    assert isinstance(p, _origen.dut.pins.Pin)
+
+  # Check initial state of pin groups
+  assert isinstance(origen.dut.pin_groups, _origen.dut.pins.PinGroupContainer)
+  assert len(origen.dut.pin_groups) == 0
+  assert origen.dut.pin_group('test_grp') is None
+  with pytest.raises(KeyError):
+    origen.dut.pin_groups['test_grp']
+
+  # Add a pin group.
+  grp = origen.dut.group_pins("test_grp", "test_pin", "other_pin")
+  assert isinstance(grp, _origen.dut.pins.PinGroup)
+
+  assert len(origen.dut.pin_groups) == 1
+  assert 'test_grp' in origen.dut.pin_groups
+  assert isinstance(origen.dut.pin_group('test_grp'), _origen.dut.pins.PinGroup)
```

Comment:
I kinda agree with @pderouen, is making the user differentiate between pins and pin_groups the right thing here?

Reply:
My main reasons were, we could return a consistent class instead of sometimes a pin and sometimes a pin collection. Also, I don't think we had an easy way to get pin groups or all available names which, from an auto-doc perspective is something we'd need. It was also a way to differentiate between a physical pin and a more abstract grouping of them.

Based on that, it seemed like just making them distinct was the way to go. We could probably meet in the middle and do something like:

~~~python
dut.add_pin('p1')
dut.add_pin('p2')
dut.group_pins('p', 'p1', 'p2')

dut.pin('p1') #=> PinGroup of size 1
dut.pin('p') #=> PinGroup of size 2

dut.pins #=> ['p1', 'p2',  'p']
dut.physical_pins #=> ['p1', 'p2']
~~~

This way, I can return a consistent class, pins/pin_groups/aliasing will all work at the `pins` level, and `physical_pins` can return the actual hardware pins. From the Rust perspective, I can keep the same scheme of the pin groups just operating on the underlying pin objects.

Reply:
@pderouen @ginty any comment here? I need to add some error handling before merging, but if the prevailing opinion is to merge pin and pin group retrieval, I need to do that first.

Reply:
I think this should be handled similar to regs. For regs most of the API will be implemented by a `BitCollection` object which can contain one or more bits. That means from a user perspective there is no API difference between dealing with a single bit, a whole register or a partial register (and hopefully we can even extend that in O2 to be larger collections like multiple registers or even whole sections of memory).

Pins should probably be the same. Internally we should have a collection of all individual pin objects and a separate collection of pin groups - which don't contain the pins themselves but just the details of what pins are included in the given group and any other metadata associated with it.
Probably another collection would be a hashmap of aliases.

`pin('blah')` should look up individual pins, alias and pin groups for a match and wrap the resolved pin(s) in a `PinCollection` for returning to the user.

Reply:
Okay. I think we're on the same page then. The only difference I have is I have a `PinGroup` as a named `PinCollection`.